docs/proposals/20220603-kubernetes-vendor.md

```diff
@@ -0,0 +1,235 @@
+
+#### envd contexts
+
+envd uses buildkitd container in the local docker host to build the images by default. `context` command will be introduced to support remote build.
+
+```
+NAME:
+   envd context - Manage envd contexts
+
+USAGE:
+   envd context command [command options] [arguments...]
+
+COMMANDS:
+   create   Create envd context
+   ls       List envd contexts
+   rm       Remove envd context
+   use      Use the specified envd context
+   help, h  Shows a list of commands or help for one command
+
+OPTIONS:
+   --help, -h  show help (default: false)
+```
+
+The structure of the envd context is:
+
+```go
+type Context struct {
+	Name          string      `json:"name,omitempty"`
+	Builder       BuilderType `json:"builder,omitempty"`
+	BuilderSocket string      `json:"builder_socket,omitempty"`
+
+	Runner     RunnerType `json:"runner,omitempty"`
+	DockerHost *string    `json:"docker_host,omitempty"`
+	KubeConfig *string    `json:"kube_config,omitempty"`
+}
+
+type BuilderType string
+
+const (
+	BuilderTypeDocker     BuilderType = "docker-container"
+	BuilderTypeKubernetes BuilderType = "kube-pod"
+)
+
+type RunnerType string
+
+const (
+	RunnerTypeDocker     RunnerType = "docker"
+	RunnerTypeKubernetes RunnerType = "kubernetes"
+)
+```
+
+The default envd context is:
+
+```go
+Context{
+    Name:          "default",
+    Builder:       BuilderTypeDocker,
+    BuilderSocket: "unix:///var/run/docker.sock",
+    Runner:        RunnerTypeDocker,
+    DockerHost:    nil,
+    KubeConfig:    nil,
+}
+```
+
+Users may use `envd context create` to create new context and use it:
+
+```
+$ envd context create --name my-context --builder-type docker-container --runner-type docker --docker-host unix:///var/run/docker.sock
+```
+
+#### Entrypoint
+
+Currently, entrypoint is set during the container creation:
+
+```go
+	config := &container.Config{
+		Entrypoint: []string{
+			"tini",
+			"--",
+			"bash",
+			"-c",
+		},
+	}
+	config.Entrypoint = append(config.Entrypoint,
+		entrypointSH(g, config.WorkingDir, sshPort))
+```
+
+To support the Kubernetes, envd should generate entrypoint and cmd for the image during the build process, instead of in the runtime.
+
+#### Bi-directional filesystem sync
+
+A per-pod sidecar is introduced to sync files between host and Kubernetes pod. [syncthing](https://docs.syncthing.net/intro/getting-started.html) will run in the per-pod sidecar.
+
+The sidecar design does not involve changing the image.
+
+#### Port forwarding
+
+There are several ports will be used in envd:
+
+- sshd server port. envd randomly selects a host port for sshd.
+- jupyter notebook port.
+- RStudio server port (in the future).
+
+[portforwarder](https://github.com/kubernetes/kubernetes/blob/v1.6.0-alpha.0/pkg/kubectl/cmd/portforward.go#L107) in client-go will be used, to forward pod ports to the host.
+
+**Pending Issues:**
+
+- `ssh envd` will be broken on Kubernetes if we do not have a daemon process to forward the ports.
```

Comment:
There are some approaches:

- Make a daemon process in the host, to forward ports
- use `envd attach` or `envd ssh` instead of `ssh`, we wrap the ssh to the correct port.